### PR TITLE
Dev audio speech emotion analyzer

### PR DESCRIPTION
**Pull request by yingzhao27**

### CODE REVIEW

File: Audio/Speech-Emotion-Analyzer/README.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+python preprocess.py --config configs/example.yaml
+```
+
+where `configs/example.yaml` is the path to your config file
+
+
+
+### Train
+
+The path of the datasets can be configured in [`config.py`](config.py). Audios which express the same emotion should be put in the same folder (you may want to refer to [`utils/files.py`](utils/files.py) when setting up datasets), for example:
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/README.md:100]: 没找到 `config.py` 这个文件

Comment (yingzhao27) [reply Audio/Speech-Emotion-Analyzer/README.md]: > 没找到 `config.py` 这个文件
已修改

File: Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    def init_weights(self):
+        stdv = 1.0 / math.sqrt(self.hidden_size)
+        for weight in self.parameters():
+            weight.data.uniform_(-stdv, stdv)
+
+    def forward(self, x, init_states=None):
+        """Assumes x is of shape (batch, sequence, feature)"""
+        seq_sz, bs, _ = x.size()
+        # print(x.size())
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py:56]: debug 的代码都删了吧

File: Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        # shape of self.hidden: (a, b), where a and b both
+        # have shape (num_layers, batch_size, hidden_dim).
+        # print('input.shape',input.shape)
+        lstm_out, _ = self.lstm(input.reshape(input.shape[0], self.batch_size, -1))
+
+        # Only take the output from the final timestep
+        # Can pass on the entirety of lstm_out to the next layer if it is a seq2seq prediction
+        # NOTE(Xu Zhiqiu) Negative indexing and view not supported
+        output = lstm_out[lstm_out.shape[0] - 1].reshape(self.batch_size, -1)
+        # print('output.shape',output.shape)
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py:33]: debug代码都删了吧

File: Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                flow.zeros((bs, self.hidden_size)).to("cuda"),
+            )
+        else:
+            h_t, c_t = init_states
+
+        HS = self.hidden_size
+        for t in range(seq_sz):
+            x_t = x[t, :, :]
+            # print('x.shape[1]', x.shape[1])
+            # print('x.shape[2]', x.shape[2])
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py:70]: debug代码删了

File: Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                flow.sigmoid(gates[:, HS * 3 :]),
+            )
+            c_t = f_t * c_t + i_t * g_t
+            h_t = o_t * flow.tanh(c_t)
+            hidden_seq.append(h_t.unsqueeze(0))
+        hidden_seq = flow.cat(hidden_seq, dim=0)
+        return hidden_seq, (h_t, c_t)
+
+
+if __name__ == "__main__":
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/models/lstm_oneflow.py:89]: main 函数不需要保留了

File: Audio/Speech-Emotion-Analyzer/models/lstm_ser.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        )
+
+    def forward(self, x):
+        # x = self.lstm_feature(x)
+
+        logits = self.classifier(x)
+        return logits
+
+
+if __name__ == "__main__":
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/models/lstm_ser.py:24]: 同上

File: Audio/Speech-Emotion-Analyzer/predict.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    logits = SER_model(test_feature)
+    result = np.argmax(logits.numpy(),)
+    print("Recognition:", config.class_labels[int(result)])
+
+    result_prob = flow.softmax(logits, dim=1)
+    utils.radar(result_prob.numpy().squeeze(), config.class_labels)
+
+
+if __name__ == "__main__":
+    audio_path = "/home/zhaoying/python_project/emotion_analysis/datasets/CASIA/angry/202-angry-wangzhe.wav"
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/predict.py:56]: 路径不要写死，换成用户可配置的，从命令行参数传入

File: Audio/Speech-Emotion-Analyzer/utils/files.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+"""Reorganizing datasets"""
+
+import os
+import shutil
+
+data_path = "/Users/zou/datasets/CASIA/6"
```
Comment (Ldpe2G) [inline Audio/Speech-Emotion-Analyzer/utils/files.py:6]: 这个变量没用到

**Comment by rentainhe** [comment]
整理一个可以一键运行的脚本:
- `train.sh`：对应README中的Train部分
- `infer.sh`：对应README中的Predict部分

删除无用的debug代码，及变量名

File: Audio/Speech-Emotion-Analyzer/train.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    train_iter = DataLoader(train_dataset, batch_size=config.batch_size, shuffle=True)
+    test_iter = DataLoader(test_dataset, batch_size=config.batch_size, shuffle=True)
+
+    # model_cnn1d = cnn1d_ser(1, config.n_kernels, n_feats, config.hidden_size, len(config.class_labels))
+    model_lstm = lstm_ser(
+        n_feats, config.rnn_size, len(config.class_labels), config.batch_size
+    )
+    loss_fn = nn.CrossEntropyLoss()
+    model_lstm.to("cuda")
+    # model_cnn1d.to("cuda")
```
Comment (rentainhe) [inline Audio/Speech-Emotion-Analyzer/train.py:67]: 所选的model是否可以通过config传入，然后在这里加一个判断，会不会灵活一些

File: Audio/Speech-Emotion-Analyzer/train.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        test_acc.append(te_acc)
+    print("Done!")
+
+    # Saving the trained model
+    model_path = os.path.join(config.checkpoint_path, config.checkpoint_name)
+    if os.path.exists(model_path):
+        shutil.rmtree(model_path)
+    flow.save(model_lstm.state_dict(), model_path)
+
+    # Visualize the training process
```
Comment (rentainhe) [inline Audio/Speech-Emotion-Analyzer/train.py:154]: Visual也可以作为一个config传入
```python
if Vis:
  draw()
```

File: Audio/Speech-Emotion-Analyzer/utils/plot.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    ax.set_thetagrids(angles * 180 / np.pi, class_labels)
+    ax.set_title("Emotion Recognition", va="bottom")
+
+    # set the maximum data value of the radar chart
+    ax.set_rlim(0, 1)
+
+    ax.grid(True)
+    # plt.ion()
+    plt.show()
+    # plt.pause(4)
```
Comment (rentainhe) [inline Audio/Speech-Emotion-Analyzer/utils/plot.py:81]: 这里一些没有用到的方法可以删除一下

File: README.md [hunk 12]
```diff
@@ -105,6 +105,7 @@ Please check our the following **demos** for a quick start
 - [SincNet](https://github.com/Oneflow-Inc/models/tree/main/Audio/SincNet)
 - [Wav2Letter](https://github.com/Oneflow-Inc/models/tree/main/Audio/Wav2Letter)
 - [AM_MobileNet1D](https://github.com/Oneflow-Inc/models/tree/main/Audio/AM-MobileNet1D)
+- [Speech-Emotion-Analyer](https://github.com/Oneflow-Inc/models/tree/dev_audio_speech_emotion_analyzer/Audio/Speech-Emotion-Analyzer)
```
Comment (Ldpe2G) [inline README.md:108]: 分支名字错了，合并之后应该是变成 main 分支

File: README_zh-CN.md [hunk 13]
```diff
@@ -106,6 +106,7 @@ OneFlow-Models目录下提供了各种经典图形分类、目标检测、图像
 - [SincNet](https://github.com/Oneflow-Inc/models/tree/main/Audio/SincNet)
 - [Wav2Letter](https://github.com/Oneflow-Inc/models/tree/main/Audio/Wav2Letter)
 - [AM_MobileNet1D](https://github.com/Oneflow-Inc/models/tree/main/Audio/AM-MobileNet1D)
+- [Speech-Emotion-Analyer](https://github.com/Oneflow-Inc/models/tree/dev_audio_speech_emotion_analyzer/Audio/Speech-Emotion-Analyzer)
```
Comment (Ldpe2G) [inline README_zh-CN.md:109]: 同上

**Review by Ldpe2G — APPROVED** [review]
